File: 0121_best_time_to_buy_and_sell_stock.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        local_min = prices[0]

        for i in range(len(prices) - 1):
            local_min = min(local_min, prices[i])
            profit = max(prices[i + 1] - local_min, profit)

        return profit

File: src/0121_best_time_to_buy_and_sell_stock/0121_best_time_to_buy_and_sell_stock.rs
use std::cmp::max;
use std::cmp::min;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut profit: i32 = 0;
        let mut local_min: i32 = prices[0];

        for i in 0..prices.len()-1 {
            local_min = min(local_min, prices[i]);
            profit = max(prices[i + 1] - local_min, profit);
        }

        return  profit;
    }
}
